File: src/main/java/club/xiaojiawei/listener/log/PowerLogListener.java
package club.xiaojiawei.listener.log;

import club.xiaojiawei.interfaces.closer.LogListenerCloser;
import club.xiaojiawei.core.Core;
import club.xiaojiawei.custom.LogRunnable;
import club.xiaojiawei.data.SpringData;
import club.xiaojiawei.enums.StepEnum;
import club.xiaojiawei.status.War;
import club.xiaojiawei.strategy.AbstractPhaseStrategy;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

import static club.xiaojiawei.enums.WarPhaseEnum.*;

/**
 * 对局日志监听器
 * @author 肖嘉威
 * @date 2023/7/5 20:40
 */
@Slf4j
@Component
public class PowerLogListener extends AbstractLogListener implements LogListenerCloser {

    @Resource
    private Core core;

    private ScheduledFuture<?> errorScheduledFuture;

    private volatile long lastWorkTime;

    private static final long MAX_IDLE_TIME = 5 * 60 * 1_000L;

    @Autowired
    public PowerLogListener(SpringData springData) {
        super(springData.getPowerLogName(), 0, 1_000L, TimeUnit.MILLISECONDS);
    }

    @Override
    protected void readOldLog() throws IOException {
        accessFile.seek(accessFile.length());
        War.reset();
    }

    @Override
    protected void listenLog() throws IOException {
        while (!isPause.get().get() && !AbstractPhaseStrategy.isDealing()){
            String line = accessFile.readLine();
            if (line == null){
                break;
            }else if (isRelevance(line)){
                resolveLog(line);
            }
        }
    }

    @Override
    protected void otherListen() {
        lastWorkTime = System.currentTimeMillis();
        log.info("开始监听异常情况");
        errorScheduledFuture = listenFileThreadPool.scheduleAtFixedRate(new LogRunnable(() -> {
            if (!isPause.get().get() && System.currentTimeMillis() - lastWorkTime > MAX_IDLE_TIME){
                log.info("监听到异常情况，准备重启游戏");
                lastWorkTime = System.currentTimeMillis();
                core.restart();
            }
        }), 0, 1, TimeUnit.MINUTES);
    }

    @Override
    protected void cancelOtherListener() {
        if (errorScheduledFuture != null && !errorScheduledFuture.isDone()){
            errorScheduledFuture.cancel(true);
        }
    }

    private void resolveLog(String line) {
        switch (War.getCurrentPhase()){
            case FILL_DECK_PHASE -> {
                War.setStartTime(System.currentTimeMillis());
                FILL_DECK_PHASE.getAbstractPhaseStrategy().deal(line);
            }
            case GAME_OVER_PHASE -> {
                War.setEndTime(War.getStartTime() == 0 ? 0 : System.currentTimeMillis());
                GAME_OVER_PHASE.getAbstractPhaseStrategy().deal(line);
                War.reset();
            }
            default -> War.getCurrentPhase().getAbstractPhaseStrategy().deal(line);
        }
        if (War.getCurrentTurnStep() == StepEnum.FINAL_GAMEOVER){
            War.setCurrentPhase(GAME_OVER_PHASE);
        }
    }

    public boolean isRelevance(String l){
        boolean flag = false;
        if (l.contains("Truncating log")){
            log.info("power.log达到10000KB，游戏停止输出日志，准备重启游戏");
            core.restart();
        }else {
            flag = l.contains("PowerTaskList");
        }
        lastWorkTime = System.currentTimeMillis();
        return flag;
    }

    @Override
    public void closeLogListener() {
        cancelListener();
    }

}

File: Hearthstone-Script-App/src/main/java/club/xiaojiawei/hsscript/listener/log/PowerLogListener.kt
package club.xiaojiawei.hsscript.listener.log

import club.xiaojiawei.config.log
import club.xiaojiawei.enums.StepEnum
import club.xiaojiawei.enums.WarPhaseEnum
import club.xiaojiawei.hsscript.bean.single.WarEx
import club.xiaojiawei.hsscript.consts.GAME_WAR_LOG_NAME
import club.xiaojiawei.hsscript.core.Core
import club.xiaojiawei.hsscript.initializer.BaseInitializer
import club.xiaojiawei.hsscript.listener.WorkTimeListener
import club.xiaojiawei.hsscript.status.PauseStatus
import club.xiaojiawei.hsscript.status.ScriptStatus
import club.xiaojiawei.hsscript.strategy.AbstractPhaseStrategy
import club.xiaojiawei.hsscript.utils.PowerLogUtil
import club.xiaojiawei.status.WAR
import java.io.RandomAccessFile
import java.util.concurrent.TimeUnit

/**
 * 对局日志监听器
 * @author 肖嘉威
 * @date 2023/7/5 20:40
 */
object PowerLogListener :
    AbstractLogListener(GAME_WAR_LOG_NAME, 0, 50L, TimeUnit.MILLISECONDS) {

    private val war = WAR

    private const val RESERVE_SIZE_B = 4 * 1024 * 1024

    override fun dealOldLog() {
        innerLogFile?.let {
            it.seek(it.length())
        }
        WarEx.reset()
    }

    override fun dealNewLog() {
        while (!PauseStatus.isPause && !AbstractPhaseStrategy.dealing && WorkTimeListener.working) {
            innerLogFile?.let {
                val line = it.readLine()
                if (line == null) {
                    return@dealNewLog
                } else if (PowerLogUtil.isRelevance(line)) {
                    resolveLog(line)
                }
            } ?: return
        }
    }

    private fun resolveLog(line: String) {
        when (war.currentPhase) {
            WarPhaseEnum.FILL_DECK -> {
                WarPhaseEnum.FILL_DECK.phaseStrategy?.deal(line)
            }

            WarPhaseEnum.GAME_OVER -> {
                WarPhaseEnum.GAME_OVER.phaseStrategy?.deal(line)
            }

            else -> war.currentPhase.phaseStrategy?.deal(line)
        }
        if (war.currentTurnStep == StepEnum.FINAL_GAMEOVER) {
            war.currentPhase = WarPhaseEnum.GAME_OVER
        }
    }

    fun checkPowerLogSize(): Boolean {
        val logFile = logFile
        logFile ?: return false

        if (logFile.length() + RESERVE_SIZE_B >= ScriptStatus.maxLogSizeB) {
            log.info { "${GAME_WAR_LOG_NAME}即将达到" + (ScriptStatus.maxLogSizeKB) + "KB，准备重启游戏" }
            Core.restart()
            return false
        }
        return true
    }

    @JvmStatic
    fun main(args: Array<String>) {
//        innerLogFile = RandomAccessFile("S:\\Hearthstone\\Logs\\Hearthstone_2024_12_04_19_06_41\\Power - 副本.log", "r")
        innerLogFile = RandomAccessFile("S:\\Hearthstone\\Logs\\Hearthstone_2024_11_27_17_38_36\\${GAME_WAR_LOG_NAME}", "r")
        WorkTimeListener.working = true
        BaseInitializer().init()
        WarEx.reset()
//        innerLogFile = RandomAccessFile("S:\\Hearthstone\\Logs\\Hearthstone_2024_12_04_19_06_41\\Power.log", "r")
        dealNewLog()
    }
}
